File: KartikJS1/microsoft/Day4/LongestHappyPrefix.cpp
class Solution {
public:
    string longestPrefix(string s) {
                int n=s.size();
        vector<int> lps(n,0);
        int i=1,len=0;
        while(i<n){
            if(s[i]==s[len]){
                len++;
                lps[i]=len;
                i++;
            }
            else{
                if(len>0){
                    len=lps[len-1];
                }
                else{
                    lps[i]=0;
                    i++;
                }
            }
        }
        return s.substr(0,lps[n-1]);

    }
};

File: Ravi/Microsoft/Day 4/LongestHappyPrefix.java
class Solution {
    public String longestPrefix(String s) {
        
        int n = s.length();
        
        int lhp[] = new int[n];
        
        int i = 1;
        int curr = 0;
        
        while(i < n) {
            
            if(s.charAt(i) == s.charAt(curr)) {
                curr++;
                lhp[i] = curr;
                i++;
            } else{
                if(curr > 0) curr = lhp[curr - 1];
                else {
                    lhp[i] = 0;
                    i++;
                }
            }
        }
        
        return s.substring(0, lhp[n - 1]);
    }
}
